### DISCUSSION TITLE
大佬，ItemsControl 模版方式 创建的自定义控件，后台代码如何处理某个满足条件的子控件(ItemControl)？

### DISCUSSION THREAD

**Discussion by rainteen** [BYJRK/DotNet-Discussions#352] [C# 技术交流]
大佬，ItemsControl 模版方式 创建的自定义控件，后台代码如何处理某个满足条件的子控件(ItemControl)？包括 选中某一项，设置选中的背景、触发相应事件等等。以及 如何遍历 ItemsControl 中所有 跟 ItemsSource对应的控件？

**Comment by rainteen** [comment]
补充一下问题。想设计一个 带标题的  单选组控件，允许绑定 特定数据列表，生成一组 单选项。xaml代码如下：
`<ResourceDictionary xmlns="http://schemas.microsoft.com/winfx/2006/xaml/presentation"
                    xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml"
                    xmlns:ctrls="clr-namespace:Lunar.Controls.Ctrls"
                    xmlns:ext="clr-namespace:Lunar.Controls.Ctrls.Ext"
                    xmlns:cvt="clr-namespace:Lunar.Controls.Ctrls.Converter">
    <SolidColorBrush x:Key="Control.Static.Foreground" Color="Black"/>
    <SolidColorBrush x:Key="Control.Static.Background" Color="Transparent"/>
    <SolidColorBrush x:Key="Control.Static.SelectionBrush" Color="#808F8787"/>

    <SolidColorBrush x:Key="Control.Static.BorderBrush" Color="#FF909090"/>
    <SolidColorBrush x:Key="Control.Focused.BorderBrush" Color="#FF007ACC"/>
    <SolidColorBrush x:Key="Control.MouseOver.BorderBrush" Color="#FF1E1E1E"/>
    <cvt:SelectedItemNullToBoolConverter x:Key="SelectedItemNullToBoolConverter"/>
    <Style x:Key="LabelRadioBox" TargetType="{x:Type ctrls:LabelRadioBox}">
        <Setter Property="Template">
            <Setter.Value>
                <ControlTemplate TargetType="{x:Type ctrls:LabelRadioBox}">
                    <Border Width="{TemplateBinding Width}" 
                            Height="{TemplateBinding Height}" 
                            MaxWidth="{TemplateBinding MaxWidth}"
                            Margin="{TemplateBinding Padding}"
                            Background="{StaticResource Control.Static.Background}"
                            BorderBrush="{StaticResource Control.Static.BorderBrush}"
                            BorderThickness="1"
                            HorizontalAlignment="{TemplateBinding HorizontalAlignment}"
                            VerticalAlignment="{TemplateBinding VerticalAlignment}">
                        <ItemsControl x:Name="PART_lrbRadioGroup" Width="{TemplateBinding Width}"
                                      ItemsSource="{Binding RelativeSource={RelativeSource AncestorType=ctrls:LabelRadioBox, Mode=FindAncestor}, Path=DataList}">
                            <ItemsControl.Template>
                                <ControlTemplate TargetType="ItemsControl">
                                    <Grid x:Name="PART_lrbGridLayout" 
                                          Margin="0" Width="{TemplateBinding Width}">
                                        <Grid.ColumnDefinitions>
                                            <ColumnDefinition Width="80"/>
                                            <ColumnDefinition Width="*"/>
                                            <ColumnDefinition Width="10"/>
                                        </Grid.ColumnDefinitions>
                                        <Label Grid.Column="0" Width="80" FontSize="16" FontWeight="Medium"
                                               Content="{Binding RelativeSource={RelativeSource AncestorType=ctrls:LabelRadioBox, Mode=FindAncestor}, Path=Header}" 
                                               BorderBrush="{StaticResource Control.Static.BorderBrush}" 
                                               BorderThickness="1"/>
                                        <UniformGrid Grid.Column="1" Columns="2" FlowDirection="LeftToRight" IsItemsHost="True">
                                        </UniformGrid>
                                        <Border Grid.Column="2" Width="10" 
                                                BorderBrush="{StaticResource Control.Static.BorderBrush}"
                                                BorderThickness="1">
                                            <TextBlock Text="*" FontWeight="Bold"
                                                       FontSize="{TemplateBinding FontSize}" 
                                                       Foreground="{StaticResource Control.Static.Background}"
                                                       Background="{StaticResource Control.Static.Background}"
                                                       TextAlignment="Center" HorizontalAlignment="Center" 
                                                       VerticalAlignment="Center">
                                                <TextBlock.Style>
                                                    <Style TargetType="TextBlock">
                                                        <Style.Triggers>
                                                            <MultiDataTrigger>
                                                                <MultiDataTrigger.Conditions>
                                                                    <Condition Property="{Binding RelativeSource={RelativeSource AncestorType=ctrls:LabelRadioBox, Mode=FindAncestor}, Path=NotEmpty}"
                                                                               Value="True"/>
                                                                    <Condition Property="{Binding RelativeSource={RelativeSource AncestorType=ctrls:LabelRadioBox, Mode=FindAncestor}, Path=SelectedItem, Converter={StaticResource SelectedItemNullToBoolConverter}}"
                                                                               Value="True"/>
                                                                </MultiDataTrigger.Conditions>
                                                                <Setter Property="Foreground" Value="Red"/>
                                                            </MultiDataTrigger>
                                                        </Style.Triggers>
                                                    </Style>
                                                </TextBlock.Style>
                                            </TextBlock>
                                        </Border>
                                    </Grid>
                                </ControlTemplate>
                            </ItemsControl.Template>
                            <ItemsControl.ItemTemplate>
                                <DataTemplate DataType="RadioButton">
                                    <RadioButton GroupName="PART_rbDataItem" IsTabStop="True" 
                                                 Content="{Binding Path=Name}" IsThreeState="False"
                                                 ext:RadioButtonAttachHelper.IsSelected="{Binding Path=IsChecked}"/>
                                    <DataTemplate.Triggers>
                                        <DataTrigger Value="True"
                                                     Binding="{Binding ElementName=PART_rbDataItem, Path=IsChecked}">
                                            <Setter Property="RadioButton.Background" Value="LightCyan"/>
                                        </DataTrigger>
                                        <DataTrigger Value="False"
                                                     Binding="{Binding ElementName=PART_rbDataItem, Path=IsChecked}">
                                            <Setter Property="RadioButton.Background" Value="Transparent"/>
                                        </DataTrigger>
                                    </DataTemplate.Triggers>
                                </DataTemplate>
                            </ItemsControl.ItemTemplate>
                        </ItemsControl>
                    </Border>
                </ControlTemplate>
            </Setter.Value>
        </Setter>
    </Style>
    <Style BasedOn="{StaticResource LabelRadioBox}" TargetType="{x:Type ctrls:LabelRadioBox}"/>
</ResourceDictionary>

`
问题，如上提问所示。

**Comment by rainteen** [comment]
`

    public class LabelRadioBox : Selector
    {
        public double SmallFontSize
        {
            get { return (double)GetValue(SmallFontSizeProperty); }
            set { SetValue(SmallFontSizeProperty, value); }
        }
        public static readonly DependencyProperty SmallFontSizeProperty =
            DependencyProperty.Register(
                "SmallFontSize",
                typeof(double),
                typeof(LabelRadioBox),
                new PropertyMetadata(0.0)
            );

        public string Header
        {
            get { return (string)GetValue(HeaderProperty); }
            set { SetValue(HeaderProperty, value); }
        }
        public static readonly DependencyProperty HeaderProperty = DependencyProperty.Register(
            "Header",
            typeof(string),
            typeof(LabelRadioBox),
            new PropertyMetadata(string.Empty)
        );

        /// <summary>
        /// 标题宽度
        /// </summary>
        public int HeaderWidth
        {
            get { return (int)GetValue(HeaderWidthProperty); }
            set { SetValue(HeaderWidthProperty, value); }
        }
        /// <summary>
        /// 标题宽度
        /// </summary>
        public static readonly DependencyProperty HeaderWidthProperty = DependencyProperty.Register(
            "HeaderWidth",
            typeof(int),
            typeof(LabelRadioBox),
            new PropertyMetadata(0)
        );

        public string GroupName
        {
            get { return (string)GetValue(GroupNameProperty); }
            set { SetValue(GroupNameProperty, value); }
        }
        public static readonly DependencyProperty GroupNameProperty = DependencyProperty.Register(
            "GroupName",
            typeof(string),
            typeof(LabelRadioBox),
            new PropertyMetadata("")
        );
        
        public List<object> DataList
        {
            get { return (List<object>)GetValue(DataListProperty); }
            set { SetValue(DataListProperty, value); }
        }
        public static readonly DependencyProperty DataListProperty = DependencyProperty.Register(
            "DataList",
            typeof(List<object>),
            typeof(LabelRadioBox),
            new PropertyMetadata(new List<object>())
        );

        static LabelRadioBox()
        {
            DefaultStyleKeyProperty.OverrideMetadata(
                typeof(LabelRadioBox),
                new FrameworkPropertyMetadata(typeof(LabelRadioBox))
            );
        }

        public LabelRadioBox()
        {
            if (System.ComponentModel.DesignerProperties.GetIsInDesignMode(this))
                return;
        }

        public override void OnApplyTemplate()
        {
            base.OnApplyTemplate();

        }
    }`

**Reply by designerlast** [reply]
既然考虑选中，为啥不用listbox而是itemscontrol呢？

**Reply by rainteen** [reply]
理论上用Selector是比较合适，主要是列表展示比较复杂，用ListBox是不是不太合适？
就比如网店商品列表之类的：
![商品展示](https://github.com/user-attachments/assets/50b8808e-be84-42c2-a7e6-bf9e58240ec3)

**Reply by BYJRK** [reply]
你的这个想法其实是错误的，你不要把 `ListBox` 想象成一个仿佛自带了你不喜欢的样式的控件。即便你用完全没有样式的 `ItemsControl`，你终究是要重写 `ItemTemplate`（以及可能还要重写 `ItemsPanel`）的。`ListBox` 是一个非常高级的 `ItemsControl`，不仅继承了 `Selector` 提供了每个项的 `IsSelected` 属性，更是自带虚拟化，能省下不少的麻烦。至于 `IsSelected`，也绝对不要把它想象成只能默认样式，选中了就背景变蓝一样，而是你想怎么重写样式都可以，Item 用 `CheckBox`，`IsSelected` 绑定 `IsChecked` 都是完全没问题的。

**Reply by rainteen** [reply]
哦哦，好的好的。非常感谢您的答复。
我好像也意识到了这个问题，ListBox和ListBoxItem展示样式也可以定制。
我试试。
向大佬学习，非常感谢。。。
